### PR TITLE
Percent Gene Coverage Task Modularization

### LINKED ISSUE #312: Percent reference coverage for Mpox genes
:cool:

### :pushpin: Explain the Request
The SARS-CoV-2 track for TheiaCoV_Illumina_PE_PHB provides an output file with percent coverage for each gene in the SARS-CoV-2 reference genome. The same functionality is being requested for the Mpox track of TheiaCoV_Illumina_PE_PHB.

### :books: Context
This functionality will make it easier to determine the location and impact of missing data in the Mpox genome without having to rely on programs external to Terra.

### :chart_with_upwards_trend: Desired Behavior

Ouput CSV file from the TheiaCoV_Illumina_PE_PHB workflow run using default settings for Mpox, containing a table with the gene names and percent reference coverage for each.

### LINKED ISSUE #325: [Gene Coverage] Modularize the gene coverage
:cool:

### :pushpin: Explain the Request

See also issue #312 

Modularize the sc2_gene_coverage task by altering it to iterate through bed files that contain regions of interest and organizing those bed files in the organism_parameter.wdl subworkflow

### PR DESCRIPTION
**Pull request by sage-wright**
This PR closes #325 and closes #312 

🗑️ This dev branch should be deleted after merging to main.

~Currently waiting on the organism parameter PR merging before adding this since it will use that workflow.~

## :brain: Aim, Context and Functionality

As TheiaCoV expands to include more organisms, having a WDL task that is hard-coded for a single organism is inefficient if we want to mimic the behavior for other organisms. This PR changes the calculation of breadth of coverage to no longer be hard-coded and is now organism-agnostic. This requires the usage of the organism_parameter logic subworklow and also enables the user to specify the particular regions they want to be listed in the output file by overwriting the default.

Default bed files are currently only provided for mpox and SC2.

## :hammer_and_wrench:  Impacted Workflows/Tasks & Changes Being Made
This will affect the behavior of the workflow(s) even if users don’t change any workflow inputs relative to the last version : Yes

- Output file now has a different disclaimer that is now applicable to all organisms, not just SC2

Running this workflow on different occasions could result in different results, e.g. due to use of a live database, "latest" docker image, or stochastic data processing : No 

#### 🔄 Data Processing 

The same calculations are used but now require a bed file as input to determine the regions of interest. This input bed file is looped through and `samtools depth` is used to determine the percentage of the sites that are above the specified minimum depth.

Docker/software or software versions changed: 

Databases or database versions changed:

Data processing/commands changed:

File processing changed:

Compute resources changed:

#### ➡️ Inputs 

New input: `reference_gene_locations_bed` which indicates that the gene locations should correspond to the same reference file that was used for alignment. 
By default, this file is provided for SC2 and mpox. The user can use this input file to overwrite the defaults.

#### ⬅️ Outputs 

The `sc2_all_genes_percent_coverage` file is now `est_percent_gene_coverage_tsv` as it is no longer SC2 specific.

#### Terra Testing

- [x] Illumina PE with mpox & sc2 [here](https://app.terra.bio/#workspaces/cdc-terra-resources/Theiagen_Wright_SC2_Sandbox/job_history/8b5cf90f-8ef0-4b81-a7f8-74e5e80d9ecb)
- [x] Illumina SE with all organisms [here](https://app.terra.bio/#workspaces/cdc-terra-resources/Theiagen_Wright_SC2_Sandbox/job_history/ee6ed87a-cd5b-4497-bb7d-c85f2b44f458)
- [x] ONT with all organisms except HIV [here](https://app.terra.bio/#workspaces/cdc-terra-resources/Theiagen_Wright_SC2_Sandbox/job_history/22c272cd-9d35-4d00-bf2d-b98f5928f882)
- [x] ClearLabs [here](https://app.terra.bio/#workspaces/cdc-terra-resources/Theiagen_Wright_SC2_Sandbox/job_history/7421c328-c36d-4660-a4e4-2c47722c8e27)

#### Suggested Scenarios for Reviewer to Test

- None of the workflows other than the ones mentioned above are actually changed. I just fixed some formatting with brackets and spacing because I saw them and needed to fix them immediately. They don't change any functionality at all.

## :microscope: Final Developer Checklist

- [x] The workflow/task has been tested locally and results, including file contents, are as anticipated
- [x] The workflow/task has been tested on Terra and results, including file contents, are as anticipated
- [x] The CI/CD has been adjusted and tests are passing (to be completed by Theiagen developer)
- [x] Code changes follow the [style guide](https://theiagen.notion.site/Style-Guide-WDL-Workflow-Development-bb456f34322d4f4db699d4029050481c)

## 🎯 Reviewer Checklist 

- [ ] All impacted workflows/tasks have been tested on Terra with a different dataset than used for development
- [ ] All reviewer-suggested scenarios have been tested and any additional
- [ ] All changed results have been confirmed to be accurate
- [ ] All workflows/tasks impacted by change/s have been tested using a standard validation dataset to ensure no unintended change of functionality
- [ ] All code adheres to the style guide
- [ ] MD5 sums have been updated
- [ ] The PR author has addressed all comments

## 🗂️ Associated Documentation (to be completed by Theiagen developer)

- [ ] Relevant documentation on the Public Health Resources "PHB Main" has been updated
- [ ] Workflow diagrams have been updated to reflect changes

### CODE REVIEW

File: tasks/quality_control/basic_statistics/task_gene_coverage.wdl [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+      depth=$(samtools depth -r "$chromosome:$start-$stop" ~{bamfile} | awk -F "\t" '{ if ($3 > ~{min_depth}) print; }' | wc -l)
+      coverage=$(python3 -c "print ( round( ($depth / ($stop - $start + 1) ) * 100, 2 ) )")
+
+      echo -e "$gene\t$coverage" >> ~{samplename}.percent_gene_coverage.tsv
+    done < ~{bedfile}
+
+  >>>
+  output {
+    Float sc2_s_gene_depth = read_float("S_GENE_DEPTH")
+    Float sc2_s_gene_percent_coverage = read_float("S_GENE_PC")
```
Comment (cimendes) [inline tasks/quality_control/basic_statistics/task_gene_coverage.wdl:62]: I'm wondering if using read_string here with an empty string as the default value wouldn't be better. I don't know how the sort would function on Terra, but at least the value wouldn't be populated in the table if it doesn't actually exist. What do you think?

**Comment by cimendes** [comment]
🥇 for adding tests and tweaks to better adhere to the style-guide!

**Comment by cimendes** [comment]
Testing MPOX and SC2 here: https://app.terra.bio/#workspaces/theiagen-validations/Theiagen_Mendes_Sandbox/job_history/b3a92d22-b8ea-4b86-be97-06f7460c8eac

**Comment by cimendes** [comment]
The two failures in ONT data are unrelated to the changes in this PR (And should be removed from the validation dataset!) https://app.terra.bio/#workspaces/cdc-terra-resources/Theiagen_Wright_SC2_Sandbox/job_history/22c272cd-9d35-4d00-bf2d-b98f5928f882

**Comment by cimendes** [comment]
⚠️ https://app.terra.bio/#workspaces/theiagen-validations/Theiagen_Mendes_Sandbox/job_history/210b6ce7-c37a-4435-a15f-8dc7c5da5d8e

The task runs successfully but the depth coverage is not being captured correctly:
image

https://job-manager.dsde-prod.broadinstitute.org/jobs/9fc8388d-4c18-4147-b5e2-55183de96ee3

File: tasks/quality_control/basic_statistics/task_gene_coverage.wdl [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    echo "Calculating gene coverage for ~{samplename} using samtools with ~{bamfile}"
+
+    samtools index ~{bamfile}
+
+    # extract chromosome name from bam file
+    chromosome=$(samtools idxstats ~{bamfile} | cut -f 1 | head -1)
+    export chromosome
+
+    if [ "~{organism}" == "sars-cov-2" ]; then
+      echo "organism sras-"
```
Comment (cimendes) [inline tasks/quality_control/basic_statistics/task_gene_coverage.wdl:28]: This echo statement needs fixing :P

**Comment by cimendes** [comment]
image

**Comment by sage-wright** [comment]
@cimendes Issue resolved!

File: tasks/quality_control/basic_statistics/task_gene_coverage.wdl (hunk 1 above)
Comment (sage-wright) [reply tasks/quality_control/basic_statistics/task_gene_coverage.wdl:61]: We'd have to change the output to a String for it to not populate to the table if it doesn't exist; doing read_float or read_string won't make much of a difference.

**Comment by cimendes** [comment]
All fixed! :D 
image

Thank you @sage-wright!

**Review by cimendes — APPROVED** [review]

**Pull request merged**